### PR TITLE
Adds data volume to lnd docker service

### PR DESCRIPTION
**Pull request by Osso**
This is my proposed changes for issue #696

### CODE REVIEW

**Comment by Osso** [comment]
This still needs the handling of lncli. It's looking for the macaroon in /root/.lnd
Somehow, this needs to be set --macaroonpath=/data/admin.macaroon

**Comment by Osso** [comment]
I added a symlink creation during the container startup to be able to use lncli again without any argument.

**Comment by dabura667** [comment]
You should never make a new folder in /

**Comment by Osso** [comment]
Actually, I did not create a new folder. The logs were already stored in /data

That said, Docker is a bit different. Since it's isolated it is quite common to store things in /<x>

**Comment by dabura667** [comment]
Why not use ~/.lnd default directory? Then you won’t need to make a symlink.

**Comment by Osso** [comment]
I find it cleaner to store data in /data and that's why I wrote it this way. If this was outside of docker I would store it in /var/lib/lnd. The home directory is not where I usually store application data.

It's definitely possible to make /root/.lnd a volume instead. I'll follow the devs opinion.

**Comment by halseth** [comment]
I think using `~/.lnd` makes the most sense, since that is what is used on Linux. Won't that be the folder used by default for docker?

**Comment by Osso** [comment]
Ok I updated the patch to make /root/.lnd a volume and removed the --datadir option

**Comment by Richard87** [comment]
What about adding `VOLUME /root/.lnd` to the lnd `dockerfile`? Then Docker would create an anonmous volume if not specified when starting a container/image... May prevent dataloss :)

**Comment by Osso** [comment]
Good idea, I added it.

**Comment by dabura667** [comment]
This should also contain a fix to the docker example readme.

It breaks the current explanation because when you try to make two lnd containers in the same host they can’t be separated, and use the same volume.

You should modify the docker/README.md to add volume parameters for alice and bob to allow separate volumes.

**Comment by Osso** [comment]
I wrote an update docker-compose.yml to add an alice and bob with different volumes.
I am not sure we want to clutter that file with services that are only useful in the tutorial.

It is possible to put them in a different docker compose file but then I need to create network for them to share.

File: docker/docker-compose.yml [hunk 1]
```diff
@@ -75,20 +75,36 @@ services:
...
 
     lnd_btc:
       extends: lnd
       container_name: lnd_btc
       links:
           - "btcd:blockchain"
+      volumes:
+          - lnd_btc_data:/root/.lnd
+
+    alice:
```
Comment (halseth) [inline docker/docker-compose.yml]: Could you make the changes to this file a separate commit that explains what they achieve?

**Comment by Osso** [comment]
I re-split the changes into 3 commits:
- the trailing spaces
- the necessary lnd volumes to prevent data loss
- the updates to make docker/readme work again

It should be more readable now

**Comment by halseth** [comment]
@Osso Are these changes/instructions up-to-date with latest master?

**Comment by halseth** [comment]
@Osso Any updates on this? Seems useful (I just cherry-picked the commit to use for some testing) 😅

File: docker/docker-compose.yml [hunk 2]
```diff
@@ -93,6 +93,19 @@ services:
       volumes:
           - lnd_btc_data:/root/.lnd
 
+    alice:
+      extends: lnd
+      links:
+          - "btcd:blockchain"
+      volumes:
+          - alice_data:/root/.lnd
```
Comment (halseth) [inline docker/docker-compose.yml]: This doesn't seem to work, as Alice an Bob will share the same folder, and won't be able to run at the same time.

Comment (Osso) [reply docker/docker-compose.yml]: alice_data and bob_data are 2 different volumes and won't conflict.

**Comment by Osso** [comment]
It's quite an old patch. I moved and did not have time to keep it up.

File: docker/docker-compose.yml (hunk 2 above)
Comment (halseth) [reply docker/docker-compose.yml]: Okay, will need to do some more testing, running into some issues... Will report back!

File: docker/README.md [hunk 3]
```diff
@@ -90,7 +90,7 @@ Connect `Bob` node to `Alice` node.
 
 ```bash
 # Run "Bob" node and log into it:
-$ docker-compose run -d --name bob lnd_btc
+$ docker-compose run -d bob
```
Comment (halseth) [inline docker/README.md]: This still needs to inlcude the `--name` parameter, since if not the container will be given a random name.

File: docker/lnd/Dockerfile [hunk 4]
```diff
@@ -34,3 +34,5 @@ RUN apk add --no-cache \
 # Copy the entrypoint script.
 COPY "docker/lnd/start-lnd.sh" .
 RUN chmod +x start-lnd.sh
+
+VOLUME /root/.lnd
```
Comment (yancyribbens) [inline docker/lnd/Dockerfile]: 👍

File: docker/docker-compose.yml [hunk 5]
```diff
@@ -82,12 +82,29 @@ services:
       container_name: lnd_ltc
       links:
           - "ltcd:blockchain"
+      volumes:
+          - lnd_ltc_data:/root/.lnd
```
Comment (yancyribbens) [inline docker/docker-compose.yml]: The volume shouldn't be required here if you've already defined it in the `Dockerfile` I believe.  Is there any reason why you're also setting it up here?

File: docker/docker-compose.yml [hunk 6]
```diff
@@ -82,12 +82,29 @@ services:
...
 
     lnd_btc:
       extends: lnd
       container_name: lnd_btc
       links:
           - "btcd:blockchain"
+      volumes:
+          - lnd_btc_data:/root/.lnd
```
Comment (yancyribbens) [inline docker/docker-compose.yml]: Same comment here.  shouldn't require it this volume if you create it in `lnd/Dockerfile`

File: docker/README.md [hunk 7]
```diff
@@ -61,19 +61,19 @@ Start `btcd`, and then create an address for `Alice` that we'll directly mine
 bitcoin into.
 ```bash
 # Init bitcoin network env variable:
-$ export NETWORK="simnet" 
+$ export NETWORK="simnet"
 
 # Run the "Alice" container and log into it:
-$ docker-compose run -d --name alice lnd_btc
+$ docker-compose run -d --name alice alice
```
Comment (yancyribbens) [inline docker/README.md]: I think it's a good idea to simplify the docs here, although it would be better to do something like:
`docker-compose exex -it --name alice` and name the container when it's run.

File: docker/docker-compose.yml (hunk 5 above)
Comment (Osso) [reply docker/docker-compose.yml]: I am mirroring the way volumes are set up for the other services.
e.g. - litecoin:/data
Also, I prefer having a named volume for ease of management.

Comment (yancyribbens) [reply docker/docker-compose.yml]: After doing some further testing it seems `docker-compose` requires this named volume for management _instead_ of an anonymous volume which is what's created in the `Dockerfile`.

File: docker/lnd/Dockerfile (hunk 4 above)
Comment (yancyribbens) [reply docker/lnd/Dockerfile]: Looks like this isn't actually needed if we use a named volume in `docker-compose`

File: docker/docker-compose.yml [hunk 8]
```diff
@@ -104,3 +121,16 @@ volumes:
...
+
+  # lnd volumes are needed for maintaining lightning channels persistance
+  # during lnd_* containers recreation.
+  lnd_ltc_data:
+     driver: local
+  lnd_btc_data:
+     driver: local
+
+  # alice and bob are need for running README examples
+  alice_data:
```
Comment (yancyribbens) [inline docker/docker-compose.yml]: IMO this `docker-compose` file should be left agnostic to implementation examples.

**Comment by guggero** [comment]
This is a very old PR and we probably should have reviewed it a long time ago.
I'm not a fan of adding tutorial specific things into the `docker-compose` file, since people might use that for their setup outside of just the tutorial.
I left some suggestions in #2533 on how to do this and am closing this PR in favor of the newer one.
